tools/autograd/templates/Functions.cpp

```diff
@@ -477,6 +477,14 @@ Tensor logsumexp_backward(Tensor grad, const Tensor & self, Tensor result, IntAr
   return grad * (self - result).exp();
 }
 
+Tensor logcumsumexp_backward(Tensor grad, const Tensor & self, Tensor result, int64_t dim) {
+  if (grad.dim() == 0 || grad.numel() == 0) {
+    return grad;
+  }
+  auto log_grad = grad * (-result).exp();
```

Comment:
The tests were passing. However I am not very knowledgeable about the same, so would be great if you or someone can let me know if there is a better way to go about it.

Reply:
Usual tests we have use values within very small range, so very unlikely to stress this.

Could you add a new test that does the following next to the one your already added please:
- Run the forward with a Tensor of input values in larger range. For example: `a = torch.rand(10); a[0] = 100`. You can check that using `torch.logsumexp()` or your functions, this will give the right result. (You can use logsumexp() on all subsets to check your function). Note that if you try to use the ground truth you currently have in the test, you will get `inf`.
- Using the same input, you can use the `gradcheck` tool from the autograd package to check that the gradients are correct:
```
# Example for logsumexp. Same can be done for your function.
a = torch.rand(10, dtype=torch.double, requires_grad=True)
with torch.no_grad():
    a[0] = 100
gradcheck(lambda x: x.logsumexp(0), a)
```

Reply:
No need to try, self.exp() is going to kill it :-)
Does tensorflow has a formula for numerically stable backward? Otherwise someone needs to derive it.

Reply:
Verified. As mentioned, it blows up when it is `1000` and above.

Digged a bit into tensorflow and it looks to have a stable implementation.

https://github.com/tensorflow/tensorflow/blob/2a5910906a0e0f3dbc186ff9db6386d81a63448c/tensorflow/python/ops/math_grad.py#L1832-L1863

Will use that as a reference.

Thanks @ngimel @albanD

Reply:
Perfect! Thanks for looking into it.
Ping me when you need a final review !

Reply:
I have tried the following which passes the gradcheck:
```python
import torch
from torch.autograd import grad
from torch.autograd import gradcheck

class LogCumsumExp(torch.autograd.Function):
    """
    We can implement our own custom autograd Functions by subclassing
    torch.autograd.Function and implementing the forward and backward passes
    which operate on Tensors.
    """

    @staticmethod
    def forward(ctx, input, dim):
        ctx.dim = dim
        with torch.no_grad():
            # Ignore Current Unstable Implementation.
            y = torch.logcumsumexp(input, ctx.dim)
        ctx.save_for_backward(input, y)
        return y

    @staticmethod
    def backward(ctx, grad_output):
        x, y = ctx.saved_tensors
        dim = ctx.dim

        grad_min = torch.tensor(torch.finfo(grad_output.dtype).min, dtype=grad_output.dtype)
        log_grad_positive = torch.where(
            grad_output > 0,
            grad_output.log(),
            grad_min)

        log_grad_negative = torch.where(
            grad_output < 0,
            (-grad_output).log(),
            grad_min)

        def reverse_logcumsum(x):
            inter = torch.logcumsumexp(torch.flip(x, (dim,)), dim)
            return torch.flip(inter, (dim,))

        output_pos = torch.exp(reverse_logcumsum(log_grad_positive - y)+ x)
        output_neg = torch.exp(reverse_logcumsum(log_grad_negative - y) + x)

        return output_pos - output_neg, None

# Input
x = torch.tensor([[10000., 0.5, 0.3], [1000., 0.5, 0.3]], dtype=torch.double)
x.requires_grad = True

logcumsumexp = LogCumsumExp()

axis = 0
gradcheck(lambda x: logcumsumexp.apply(x, axis), (x, ))

axis = 1
gradcheck(lambda x: logcumsumexp.apply(x, axis), (x, ))
```

However I have a doubt on how to go about 
```
grad_min = torch.tensor(torch.finfo(grad_output.dtype).min, dtype=grad_output.dtype)
```
in the C++ Gradient code.

Thanks.

Reply:
std::numeric_limits<scalar_t> would give you same constants that torch.finfo does.

Reply:
Thanks for the quick reply.
Where can I get the `scalar_t` or would we need to add a dispatch based on `scalar_t` for double and float?